src/test/java/org/assertj/core/api/BaseAssertionsTest.java

```diff
@@ -67,15 +69,16 @@ public int compare(Method method1, Method method2) {
         // static vs not static is not important Soft vs Not Soft assertions
         boolean equal = (ACCESS_MODIFIERS & method1.getModifiers() & method2.getModifiers()) != 0;
         equal = equal && (ignoreDeclaringClass || method1.getDeclaringClass().equals(method2.getDeclaringClass()));
-        equal = equal && (ignoreReturnType || method1.getReturnType().equals(method2.getReturnType()));
+        equal = equal && (ignoreReturnType || canonize(method1.getGenericReturnType())
+          .equals(canonize(method2.getGenericReturnType())));
         equal = equal && (ignoreMethodName || method1.getName().equals(method2.getName()));
```

Comment:
for better readability, I would prefer to extract each return type comparison to a private static method, something like:
```
private boolean sameReturnType(Method method1, Method method2) {
  return canonize(method1.getGenericReturnType()).equals(canonize(method2.getGenericReturnType()));
}
```
I believe the logical condition will be more expressive: `ignoreReturnType || sameReturnType(method1, method2))`.
obviously we should do that we the other conditions.